OpenRA.Game/Graphics/TerrainSpriteLayer.cs

```diff
@@ -25,42 +25,70 @@ public sealed class TerrainSpriteLayer : IDisposable
 		readonly Sheet[] sheets;
 		readonly Sprite emptySprite;
 
+		readonly int binSize;
+		readonly int binCols;
+		readonly int binStride;
+
```

Comment:
I suspect that one bin per cell would dramatically slow down flushing changes to the GPU. However, you do raise a good point that I haven't shown that the 2d bins are actually better than just using the new buffer updating/drawing changes to optimize drawing partial rows. It may be that increasing the number of `DrawPrimitives` calls by a factor of 5 ends up being neglible compared to the extra work involved in indexing by bin. Need to profile this.